sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ListStateImpl.scala

```diff
@@ -103,12 +108,14 @@ class ListStateImpl[S](
      newState.foreach { v =>
        val encodedValue = stateTypesEncoder.encodeValue(v)
        store.merge(encodedKey, encodedValue, stateName)
+       TWSMetricsUtils.incrementMetric(metrics, "numUpdatedStateRows")
      }
    }
 
    /** Remove this state. */
    override def clear(): Unit = {
      store.remove(stateTypesEncoder.encodeGroupingKey(), stateName)
+     TWSMetricsUtils.incrementMetric(metrics, "numRemovedStateRows")
```

Comment:
Hm, every time we append a value we add one to the `numUpdatedStateRows` metric, even though we're not actually adding a _row_ to RocksDB, we're merging something into an existing row.

But over here, we're counting removing an _entire_ row (which may have multiple values all merged together) as one removal. So then if you do `k` calls to `appendValue` and then a `clear()`, `numUpdatedRows != numRemovedStateRows`.

I don't have a good solution for you (without iterating through everything in `clear()`), but this does seem trickier than it might have initially seemed.

Reply:
Maybe it's fine to just get a `valuesIterator` when you `clear()`, and then in the future, we can implement this list as a linked list.

Reply:
@neilramaswamy - yes thats correct. we have that inherent limitation here since we either need to iterate on all the values on clear just to get the count or if we want to count only once - we need to do a `get` explicitly each time we call `appendValue`. Also, we don't really expect users to tally these numbers across batches since these are cumulative across all state operations on all variables. If you still think its too confusing for the users, I think its prob better to just mark this as `TODO` for listState and we will circle back to this issue